### PR TITLE
(Feature) User should have his phone verified via sms

### PR DESCRIPTION
**Pull request by fernandomg**
[#163177065](https://www.pivotaltracker.com/story/show/163177065)

Created two endpoints:
 - `/user/mobile` to register and send OTP to user's mobile
 - `/validate/mobile` to process the OTP entered by the user, and validate its phone number

If in `development` mode, the app won't send any SMS it will just accept as valid the requests.

### CODE REVIEW

File: src/server/verification/verifications.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class Verifications {
+  log: any
+
+  constructor() {
+    this.log = logger.child({ from: 'Verifications' })
+  }
+
+  verifyUser = async (user: UserRecord, verificationData: any): Promise<boolean> => Promise.resolve(true)
+
+  verifyMobile = async (user: UserRecord, verificationData: any): Promise<any> => {
```
Comment (sirpy) [inline src/server/verification/verifications.js]: You can add type to the verification data for mobil ie {otp:string}

File: src/imports/types.js [hunk 2]
```diff
@@ -10,9 +10,11 @@ export type UserRecord = {
 export interface StorageAPI {
     addUser(user: UserRecord): Promise<boolean>,
     updateUser(user: UserRecord): Promise<boolean>,
-    deleteUser(user: UserRecord): Promise<boolean>
+    deleteUser(user: UserRecord): Promise<boolean>,
+    storeOTP(user: UserRecord, otp: number): Promise<boolean>
```
Comment (sirpy) [inline src/imports/types.js]: I don't think we need storeOTP/getOTP, simpler to just use updateUser and add the otp details to the UserRecord.
the UserRecord is available in express requests.

File: src/server/server-middlewares.js [hunk 3]
```diff
@@ -3,23 +3,43 @@ import { Router } from 'express'
...
 
-import { GunDBPrivate } from "./gun/gun-middleware"
-import logger from "../imports/pino-logger"
+import logger from '../imports/pino-logger'
+import conf from './server.config'
 
 function wrapAsync(fn: Function) {
-  return function (req: $Request, res: $Response, next: NextFunction) {
+  return function (req: $Request & { log: any }, res: $Response, next: NextFunction) {
+    const log = req.log.child({ from: 'wrapAsync' })
```
Comment (sirpy) [inline src/server/server-middlewares.js]: Here maybe its ok to use the express logger, to show each request details.
But in general inside the handlers themselves the logger from the "req" is too much information, lets use the regular pino logger inside the handlers

File: src/server/storage/storageAPI.js [hunk 4]
```diff
@@ -3,20 +3,29 @@ import { Router } from 'express'
...
   app.post("/user/delete", passport.authenticate("jwt", { session: false }), wrapAsync(async (req, res, next) => {
-    const { user, body } = req
-    await storage.deleteUser(body.user)
+    const { body: { user } } = req
+    await storage.deleteUser(user)
     res.json({ ok: 1 })
   }));
-}
 
+  app.post("/user/mobile", passport.authenticate("jwt", { session: false }), onlyInProduction, wrapAsync(async (req, res, next) => {
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: move this endpoint also to the verification. and use storage.updateUser (as commented before)
verify/sendotp

File: src/server/verification/verificationAPI.js [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
   }))
 
+  app.post('/verify/mobile', passport.authenticate('jwt', { session: false }), onlyInProduction, wrapAsync(async (req, res, next) => {
+    const { user, body: { verificationData }, log: logger }: { user: UserRecord, body: any, log: any } = req
+    const log = logger.child({ from: 'verificationAPI - verify/mobile' })
+
+    log.debug('mobile verified', user, verificationData)
+
+    await verifier.verifyMobile(user, verificationData)
+    await GunDBPrivate.deleteOTP(user)
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: use storage.updateUser

File: src/imports/types.js (hunk 2 above)
Comment (fernandomg) [reply src/imports/types.js]: The idea of having a different gundb document was to not populate the UserRecord with temp data. But that data can be optional in the UserRecord and can be removed after confirming. So I'll change it. Thanks!

File: src/server/server-middlewares.js (hunk 3 above)
Comment (fernandomg) [reply src/server/server-middlewares.js]: Oh, sure... temp logging committed, sorry.

It was useful for me to log the errors as, sometimes, I got a simple `ERROR {}` logged and this helped me a lot to easily find them and committed.

Comment (sirpy) [reply src/server/server-middlewares.js]: i'm not referring to the temp logging, the pino-express extension adds a logger to the "req" object. this logger logs lotsa info about the http request which we don't need for every log.debug, so its better to use a per file logger "import logger from './logger/pino-logger'"

Comment (fernandomg) [reply src/server/server-middlewares.js]: Oh, if that's the issue we can just update the pino options and remove `req`, `res` from logging.

image

**Comment by fernandomg** [comment]
@sirpy, changes applied

**Pull request merged**